src/apps/metadata/avus.clj

```diff
@@ -5,22 +5,43 @@
             [apps.service.apps.de.validation :as validation]
             [cheshire.core :as json]))
 
+(defn- load-protected-avus
+  [username app-id]
+  (->> (:avus (metadata-client/list-avus username app-id))
+       (filter (comp validation/protected-attrs :attr))))
+
+(defn- remove-protected-avus
+  [avus admin?]
+  (if admin? avus (remove (comp validation/protected-attrs :attr) avus)))
+
 (defn list-avus
+  "Lists AVUs associated with an app. If `admin?` is `true` then all attributes will be listed. If `admin?`
+   is `false` then protected attributes will not be included in the listing."
   [{username :shortUsername :as user} app-id admin?]
   (let [app (app-db/get-app app-id)]
     (validation/verify-app-permission user app "read" admin?)
-    (metadata-client/list-avus username app-id)))
+    (update (metadata-client/list-avus username app-id) :avus remove-protected-avus admin?)))
 
 (defn set-avus
-  [{username :shortUsername :as user} app-id request admin?]
+  "Sets AVUs on an app. If `admin?` is `true` then any attribute may be updated and all attributes on the app
+   will be replaced. If `admin?` is `false` then protected attributes may not be included in the request body,
+   and any protected attributes that are already associated with the app will remain associated with the app."
+  [{username :shortUsername :as user} app-id {:keys [avus]} admin?]
   (let [{app-name :name :as app} (app-db/get-app app-id)]
     (validation/verify-app-permission user app "write" admin?)
-    (categorization/validate-app-name-in-hierarchy-avus username app-id app-name (:avus request))
-    (metadata-client/set-avus username app-id (json/encode request))))
+    (validation/validate-attrs-not-protected admin? avus)
+    (categorization/validate-app-name-in-hierarchy-avus username app-id app-name avus)
+    (let [new-avus (if admin? avus (concat avus (load-protected-avus username app-id)))]
+      (update (metadata-client/set-avus username app-id (json/encode {:avus new-avus}))
+              :avus remove-protected-avus admin?))))
```

Comment:
Nice 👍 

This is good as-is, but for some reason I got tripped up for a second when I initially thought the `update` was further modifying the request getting sent to the metadata service, but I realize now it's modifying the response before passing it back to the client. Not sure if `as->` threading might help clarify the flow here 🤷 But maybe my :clojure: 👀 are just a bit rusty these days 😆

Reply:
Hmm, yeah, a threading macro might help here. Maybe something like this would be easier to read.

``` clojure
(-> (metadata-client/set-avus username app-id (json/encode {:avus new-avus}))
    (update :avus remove-protected-avus admin?))
```

Reply:
Or the threading macro option that you mentioned:

``` clojure
(as-> (json/encode {:avus new-avus}) r
      (metadata-client/set-avus username app-id r)
      (update r :avus remove-protected-avus admin?))
```

Reply:
Cool, I'll go with that one.

Reply:
Yep, I was thinking of something more broken up like the second option you posted 👍